### DISCUSSION TITLE
Can pull requests change actions?

### DISCUSSION THREAD

**Discussion by dominickpastore** [community/community#27084] [Actions]
<p>One of the most common use cases for GitHub Actions is to run a test suite against pull requests.</p>
<p>What happens if the pull request includes changed actions? I’m guessing the changed actions won’t run on the pull request itself (unlike actions changed in a push, which <em>do</em> run on the commit they are part of), otherwise that seems like a big security concern. Someone could use that to steal your secrets, for example.</p>
<p>That said, I haven’t been able to find confirmation of this. Does anyone know for sure?</p>

**Comment by Simran-B** [comment, marked as answer]
<p>Pull requests can have their branch either in the source repo (yours) or a forked repo.</p>
<ul>
<li>
<p>Only users with write access to your repo can push to and create new branches in your repo, so it is assumed that they are trustworthy.</p>
</li>
<li>
<p>If the head branch is in a forked repo (external contributor), there is no access to secrets in the workflow. They will simply be not set. This applies to <code>on: pull_request</code></p>
</li>
<li>
<p>With <code>on: pull_request_target</code> as trigger, secrets can be accessed, but the workflow of the base branch is used (so the version trusted by you). Any changes done to <code>.github/workflows/</code> by the PR are ignored, which should prevent malicious users from leaking secrets. Note that also the code of the base branch is used unless you explicitly checkout the PR code:</p>
<pre><code class="lang-yaml"> - uses: actions/checkout@v2
   with:
     ref: ${{ github.event.pull_request.head.ref }}
     repository: ${{ github.event.pull_request.head.repo.full_name }}
</code></pre>
</li>
</ul>
<p>If you use actions owned by someone else and pass your secrets to them, then you need to make sure that they don’t misuse them. As such actions could be modified to be malicious at any time, you may want to review their code and specify the exact commit hash in your workflow instead of a branch or (partial) version tag:</p>
<pre><code class="lang-yaml">- uses: somebodys/action@c4fb90b
</code></pre>
<p>Also see the documentation:</p>
<ul>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsuses">jobs.&lt;job_id&gt;.steps.uses</a></li>
<li><a href="https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions">Using third-party actions</a></li>
</ul>
<p>You can also opt-in to only allow certain actions: <a href="https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-organizations-and-teams/disabling-or-limiting-github-actions-for-your-organization#allowing-specific-actions-to-run">Allowing specific actions to run</a></p>
<p>I’m not sure whether local actions could pose a problem, in particular actions which you maintain in the same repo in combination with <code>pull_request_target</code>. The checkout code I showed above would already exist and get the code from the forked repo, including modified actions.They could then access secrets, unless there is a protection against altered local actions similar to how workflow changes are ignored when <code>pull_request_target</code> is used. This is worth to investigate.</p>

**Comment by dominickpastore** [comment]
<p>Thank you for a very thorough response!</p>

**Comment by Simran-B** [comment]
<p>Just to make it clear if it wasn’t: Using <code>pull_request_target</code> and explicitly checking out the head branch is not safe, as everyone will be able to change the workflow to leak secrets:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/7819991/181101613-169d8a77-7a14-47ea-8967-208f00ab2d0a.png" class="site-icon" width="72" height="72">

      <a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc" title="12:00AM - 15 December 2020">GitHub Security Lab – 15 Dec 20</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/7819991/181101615-ac0f1923-9aca-4f5a-ba50-cf90981f3e80.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_10x10.png"></div>

<h3><a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc">Keeping your GitHub Actions and workflows secure: Preventing pwn requests</a></h3>

  <p>In this article, we’ll discuss some common security malpractices for GitHub Actions and workflows, and how to best avoid them. Our examples are based on real-world GitHub workflow implementation vulnerabilities the GitHub Security Lab has reported to...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
